### PR TITLE
Iterator to make it easier to work with a window of blocks in the RAPIDS shuffle

### PR DESCRIPTION
**Pull request by abellina**
Signed-off-by: Alessandro Bellina <abellina@nvidia.com>

This introduces an iterator that will be used in subsequent commits for the rapids shuffle. Adding it separate from them to reduce the size of PRs. If reviewers want me to put up a bigger PR with the rest of this I can, but I am trying to keep it reviewable.

The iterator, given a set of blocks will return on `next()` ranges of blocks that pack a window (of bytes)

1. If you have blocks smaller than a window, this puts them back to back within such window (spark's overpartition case)
2. If you have blocks larger than a window, this works to split ranges given the window length (multiple calls to next).

The main advantage of the approach supported by this iterator is that we utilize bounce buffers more efficiently, and also reduce the amount of calls to the underlying protocol.

### CODE REVIEW

**Comment by revans2** [comment]
The code looks fine, I just don't like to add in code if it is not actively being used.

**Comment by abellina** [comment]
@revans2 that's ok, I am fine closing this and trying to fit it in with another PR that targets mostly the receive/send states in the rapids shuffle. It's so far a pretty big diff 1.5K lines, so I can try reducing it.

**Comment by revans2** [comment]
I am fine with putting this in if I know that there will be something coming shortly.  So if you could put up a follow on PR that depends on this, then I would feel better about it, because then it is clear that you are breaking it up and that there will not be dead code for long.

**Comment by abellina** [comment]
@revans2 marked it as WIP, as I put up the rest. It is going to take me a couple of days to get that second PR up likely.

**Comment by abellina** [comment]
@jlowe @revans2 this is one is ready to review. I have my bigger PR ready to go as well on top of this, so we'll have that one up shortly after/if this one gets merged.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+/**
+ * Specifies a start and end range of byes for a block.
+ * @param block - a BlockWithSize instance
+ * @param rangeStart - byte offset for the start of the range
+ * @param rangeEnd - byte offset for the end of the range
+ * @tparam T - the specific type of `BlockWithSize`
+ */
+case class BlockRange[T <: BlockWithSize](
+    block: T, rangeStart: Long, rangeEnd: Long) {
+
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala:40]: Do we want an assertion to verify that start < end or does it even matter?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  def rangeSize(): Long = rangeEnd - rangeStart + 1
+
+  def isComplete(): Boolean = rangeEnd == block.size - 1
+}
+
+/**
+ * Given a set of blocks, this iterator returns BlockRanges
+ * of such blocks that fit `windowSize`.
+ *
+ * If a block is too large for the window, the block will be
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala:54]: I understand what it is doing, but it is not 100% clear from just the javadocs.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala (hunk 1 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala:40]: Can't hurt to add for sure. Note that these are inclusive, if you want a 1 byte window:
```
BlockRange(block, 0, 0)
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala (hunk 2 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala:54]: Ok, let me rephrase the docs.

**Comment by abellina** [comment]
@revans2 your comments should (hopefully) be addressed in: https://github.com/NVIDIA/spark-rapids/pull/934/commits/8bd887a04f9ee94e8ffaa4b37d2fcfa39b780394. Just let me know.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala (hunk 1 above)
Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala:42]: The scaladoc should explain that these are both inclusive.  By the way. I find that a bit odd.  Almost everything uses the convention of start=inclusive and end=exclusive, so this will be confusing.  Is there a compelling reason to keep the range end inclusive?  Changing this would also clean up all the +1 / -1 stuff going on which seems like it's begging for an off-by-one error to show up at some point down the road.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+ * @tparam T - the specific type of `BlockWithSize`
+ * @note this class does not own `transferBlocks`
+ * @note this class is not thread safe
+ */
+class WindowedBlockIterator[T <: BlockWithSize](blocks: Seq[T], windowSize: Long)
+    extends Iterator[Seq[BlockRange[T]]] {
+
+  require(windowSize > 0, s"Invalid window size specified $windowSize")
+
+  case class BlockWindow(start: Long, size: Long) {
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala:93]:
```suggestion
  private case class BlockWindow(start: Long, size: Long) {
```

Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala:93]: This comment applies to the other case classes that are specific to this class.

Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala:93]: Agree. `rangeEnd` is actually quite private. In all other code external to the file I end up using `rangeSize`. I'll remove -1's, and +1's.

Will also fix the access of `BlockWindow`

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+  private[this] var window = BlockWindow(0, windowSize)
+  private[this] var done = false
+
+  // helper class that captures the start/end byte offset
+  // for `block` on creation
+  case class BlockWithOffset[T <: BlockWithSize](
+      block: T, startOffset: Long, endOffset: Long)
+
+  private var lastOffset = 0L
+  private[this] val blocksWithOffsets = blocks.map { block =>
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala:110]: lastOffset doesn't need to be a class member since it's only needed to initialize `blocksWithOffsets`.  Can update to something like this:
```scala
private[this] val blocksWithOffsets = {
  var lastOffset = 0L
  blocks.map { block =>
   ...
  }
}
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala (hunk 1 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala:42]: `end` is now exclusive. `BlockRange(block, 0, 1)` is how one would specify a 1-byte block.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala (hunk 4 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala]: done.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala (hunk 3 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala]: done.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+      }
+      thisBlock = thisBlock + 1
+    }
+    val lastBlock = blockRangesInWindow.last
+    BlocksForWindow(lastBlockIndex,
+      blockRangesInWindow,
+      !continue || !lastBlock.isComplete())
+  }
+
+  private var lastSeenBlock = 0
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala:157]: Nit: Would like to see this at the top of the class next to the other member variables and marked `private[this]` for consistency with those.

**Review by jlowe — DISMISSED** [review]
Looks good to me, just a minor nit.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala (hunk 5 above)
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala]: Sure thing: https://github.com/NVIDIA/spark-rapids/pull/934/commits/9a5c6d790361a25f4f9e4f146803975287f522df

**Comment by abellina** [comment]
build

**Review by jlowe — DISMISSED** [review]

**Comment by abellina** [comment]
build

**Review by jlowe — APPROVED** [review]

**Review by revans2 — APPROVED** [review]

**Comment by abellina** [comment]
I believe blossom-ci is stuck, something about a volume having issues. I can try and kill the job and star CI over.

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
build

**Comment by abellina** [comment]
Rebased to pick up CI fix for 3.1.0

**Review by jlowe — APPROVED** [review]

**Comment by abellina** [comment]
build
